### ISSUE TITLE
Quick tip on how to use multiple GPU's

### ISSUE THREAD

**Issue by youmebangbang** [CompVis/stable-diffusion#171]
If you set your CUDA_VISIBLE_DEVICES env variable in the shell before running one of the scripts you can choose which GPU it will run on.  Then you can have multiple sessions running at once.

In windows: set CUDA_VISIBLE_DEVICES=[gpu number, 0 is first gpu]
In linux: export CUDA_VISIBLE_DEVICES=[gpu number]

**Comment by TaiPhamD** [comment]
can you use all 3 GPU to accelerate 1 session?

**Comment by youmebangbang** [comment]
I don't think distributed inference/training has been implemented yet. Maybe with the diffusers library (i don't use)?

**Comment by Daryl149** [comment]
Cool, was looking into this, felt like a waste of gpu number 2. Might work on exposing this as a configurable argument.
Also, if you are blessed with a high VRAM gpu, just opening another terminal will also use another session on the same device. So if your current usage is 10GB and you have an RTX 3090 (total of 24GB VRAM), you can open a new terminal that uses another 10GB of the remaining VRAM.

**Comment by MSDNAndi** [comment]
There are probably multiple ways to achieve that. The most straightforward one is likely to run each iteration on one GPU so when doing multiple iterations it would be accelerated but not when increasing batch size only.

**Comment by gilarto** [comment]
Having a similar problem. Visual artist here, I know next to nothing about python or coding, please bear with me.

Attempting to do a local SD 1.5 Install

Alienware R417 Gaming Laptop, dual GPU config 

(Intel HD Graphics 630, 

NVDIA GeForce GTX 1070)
w/8 GB VRAM, CUDA, Pascal Architecture)

Followed this tutorial and (apparently) managed to succesfully install the 1.5 model from Hugging Face & the AUTOMATIC1111 Web UI

https://youtu.be/ycQJDJ-qNI8

Unfortunately. when running the WebUI.bat I keep getting the following error:

`RuntimeError: [enforce fail at ..\c10\core\impl\alloc_cpu.cpp:81] data. DefaultCPUAllocator: not enough memory: you tried to allocate 12582912 bytes.`

I can't get past that stage, unable to even launch the Stable Diffusion ui on my browser.

Any ideas? I watched a couple of videos about some resource called xformers, looks like a possible solution for the low VRAM issue but honestly I'm running blind... I have a very very vague idea of what I'm doing here :-/

**Comment by krakensurf** [comment]
RE: "I can't get past that stage, unable to even launch the Stable Diffusion ui on my browser." - sounds like you haven't installed all of the dependencies required.  Try reinstalling and watch the CMD prompt closely for other errors that might fly by.   Also, make sure you run CMD as "Administrator" when launching, I had all sorts of permission errors if I try to launch it as basic user.

**Comment by krakensurf** [comment]
> If you set your CUDA_VISIBLE_DEVICES env variable in the shell before running one of the scripts you can choose which GPU it will run on. Then you can have multiple sessions running at once.
> 
> In windows: set CUDA_VISIBLE_DEVICES=[gpu number, 0 is first gpu] In linux: export CUDA_VISIBLE_DEVICES=[gpu number]

I've found numerous references in the code that indicates there is the "awareness" of multiple GPU's.  Such as:
`args.num_gpus = torch.cuda.device_count()`

But this interests me the most and plan to try it out. :
```
  # if n_gpu is > 1 we'll use nn.DataParallel.
	Line 101:             # If you only want to use a specific subset of GPUs use `CUDA_VISIBLE_DEVICES=0`
	Line 104:             # GPUs available in the environment, so **`CUDA_VISIBLE_DEVICES=1,2`** with `cuda:0`
	Line 105:             # will use the first GPU in that env, i.e. GPU#1
	Line 109:             self._n_gpu = torch.cuda.device_count()
```

**Comment by gilarto** [comment]
> RE: "I can't get past that stage, unable to even launch the Stable Diffusion ui on my browser." - sounds like you haven't installed all of the dependencies required. Try reinstalling and watch the CMD prompt closely for other errors that might fly by. Also, make sure you run CMD as "Administrator" when launching, I had all sorts of permission errors if I try to launch it as basic user.

Will do, I'll give it a shot and get back to you ASAP. Thanks so much.

**Comment by gilarto** [comment]
> > If you set your CUDA_VISIBLE_DEVICES env variable in the shell before running one of the scripts you can choose which GPU it will run on. Then you can have multiple sessions running at once.
> > In windows: set CUDA_VISIBLE_DEVICES=[gpu number, 0 is first gpu] In linux: export CUDA_VISIBLE_DEVICES=[gpu number]
> 
> I've found numerous references in the code that indicates there is the "awareness" of multiple GPU's. Such as: `args.num_gpus = torch.cuda.device_count()`
> 
> But this interests me the most and plan to try it out. :
> 
> ```
>   # if n_gpu is > 1 we'll use nn.DataParallel.
> 	Line 101:             # If you only want to use a specific subset of GPUs use `CUDA_VISIBLE_DEVICES=0`
> 	Line 104:             # GPUs available in the environment, so **`CUDA_VISIBLE_DEVICES=1,2`** with `cuda:0`
> 	Line 105:             # will use the first GPU in that env, i.e. GPU#1
> 	Line 109:             self._n_gpu = torch.cuda.device_count()
> ```

Yeah I tried that suggestion from another thread but it didn't solve the issue, WebUI launch stuck on the Runtime Error I quoted above :-(

**Comment by MSDNAndi** [comment]
> > > If you set your CUDA_VISIBLE_DEVICES env variable in the shell before running one of the scripts you can choose which GPU it will run on. Then you can have multiple sessions running at once.
> > > In windows: set CUDA_VISIBLE_DEVICES=[gpu number, 0 is first gpu] In linux: export CUDA_VISIBLE_DEVICES=[gpu number]
> > 
> > 
> > I've found numerous references in the code that indicates there is the "awareness" of multiple GPU's. Such as: `args.num_gpus = torch.cuda.device_count()`
> > But this interests me the most and plan to try it out. :
> > ```
> >   # if n_gpu is > 1 we'll use nn.DataParallel.
> > 	Line 101:             # If you only want to use a specific subset of GPUs use `CUDA_VISIBLE_DEVICES=0`
> > 	Line 104:             # GPUs available in the environment, so **`CUDA_VISIBLE_DEVICES=1,2`** with `cuda:0`
> > 	Line 105:             # will use the first GPU in that env, i.e. GPU#1
> > 	Line 109:             self._n_gpu = torch.cuda.device_count()
> > ```
> 
> Yeah I tried that suggestion from another thread but it didn't solve the issue, WebUI launch stuck on the Runtime Error I quoted above :-(

your issue has nothing to do with multiple GPUs.
I suggest to follow the instructions from the repo
I would use conda (miniconda/anaconda). best to install that before getting started.
https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Install-and-Run-on-NVidia-GPUs
Then check the troubleshooting section
https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Troubleshooting
https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Command-Line-Arguments-and-Settings

which of the settings did you try?
try --lowvram

**Comment by krakensurf** [comment]
I tried 'CUDA_VISIBLE_DEVICES=0,1' and it only used "1" instead of both.  However, other documents nVidia say it should have worked.  Maybe integrating the value with something other than CMD is valid.  
[](https://developer.nvidia.com/blog/cuda-pro-tip-control-gpu-visibility-cuda_visible_devices/)
